#### 1. Single end 

- Errors: probabily due to -X 2 set in fastq-dump 

```

(snakemake_sra) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_sra$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	1	get_fastq
	2	rename
	4
Select jobs to execute...

[Wed Apr 14 20:04:13 2021]
rule get_fastq:
    output: fastq/SRR13508743_1.fastq.gz, fastq/SRR13508748_1.fastq.gz
    jobid: 2

Job counts:
	count	jobs
	1	get_fastq
	1
Rejected 2 READS because READLEN < 1
Read 2 spots for SRR13508743
Written 2 spots for SRR13508743
Rejected 2 READS because READLEN < 1
Read 2 spots for SRR13508748
Written 2 spots for SRR13508748
[Wed Apr 14 20:04:19 2021]
Finished job 2.
1 of 4 steps (25%) done
Select jobs to execute...

[Wed Apr 14 20:04:19 2021]
rule rename:
    input: fastq/SRR13508743_1.fastq.gz, fastq/SRR13508748_1.fastq.gz
    output: fastq/Treatment_1.fastq.gz
    jobid: 3
    wildcards: out=Treatment_1

[Wed Apr 14 20:04:19 2021]
rule rename:
    input: fastq/SRR13508743_1.fastq.gz, fastq/SRR13508748_1.fastq.gz
    output: fastq/Control_1.fastq.gz
    jobid: 1
    wildcards: out=Control_1

Job counts:
	count	jobs
	1	rename
	1
Job counts:
	count	jobs
	1	rename
	1
mv: cannot stat 'fastq/SRR13508743_1.fastq.gz': No such file or directory
[Wed Apr 14 20:04:19 2021]
Error in rule rename:
    jobid: 0
    output: fastq/Control_1.fastq.gz

RuleException:
CalledProcessError in line 48 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:
Command 'set -euo pipefail;  mv fastq/SRR13508743_1.fastq.gz fastq/Control_1.fastq.gz' returned non-zero exit status 1.
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile", line 48, in __rule_rename
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
[Wed Apr 14 20:04:19 2021]
Finished job 3.
2 of 4 steps (50%) done
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_sra/.snakemake/log/2021-04-14T200412.970230.snakemake.log
```


- Snakefile

```
# This workflow is designed to download fastq files from SRA database. 
# It's possible to perform manually as well (see https://github.com/Mira0507/using_SRA)

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail; source ~/.bashrc; ')
shell.executable('/bin/bash')

rule all: 
    input: 
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])


rule get_fastq: 
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output: 
        expand("fastq/{srr}_{end}.fastq.gz", srr=config['SRA'], end=config['END'])
    params:
        SRA=config['SRA']
    run:
        shell("set +o pipefail; "
              "fastq-dump --split-files {params.SRA} --gzip -O fastq -X 2")

rule rename: 
    """
    This rule renames fastq.gz files
    """
    output: 
        "fastq/{out}.fastq.gz"
    input: 
        expand("fastq/{srr}_{end}.fastq.gz", srr=config['SRA'], end=config['END'])
    params: 
        dic=config['SAMPLE'],
        pairs=config['END']
    run:
        for key, value in params.dic.items(): 
            shell("mv fastq/{key}_1.fastq.gz fastq/{value}_1.fastq.gz")
            if len(params.pairs) == 2: 
                shell("mv fastq/{key}_2.fastq.gz fastq/{value}_2.fastq.gz")
```

- config_single.yaml

```

SAMPLE:
  SRR13508743: 'Control'
  SRR13508748: 'Treatment'

SRA:
  - 'SRR13508743'
  - 'SRR13508748'

OUTPUT_LIST:
  - 'Control_1'
  - 'Treatment_1'

END: 
  - '1'



```


#### 1. Single end 

- Errors: 


```

(snakemake_sra) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_sra$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	1	get_fastq
	2	rename
	4
Select jobs to execute...

[Wed Apr 14 23:54:36 2021]
rule get_fastq:
    output: fastq/SRR13508743_1.fastq.gz, fastq/SRR13508748_1.fastq.gz
    jobid: 2

Job counts:
	count	jobs
	1	get_fastq
	1
Rejected 35871352 READS because READLEN < 1
Read 35871352 spots for SRR13508743
Written 35871352 spots for SRR13508743
Rejected 46363423 READS because READLEN < 1
Read 46363423 spots for SRR13508748
Written 46363423 spots for SRR13508748
[Thu Apr 15 00:21:15 2021]
Finished job 2.
1 of 4 steps (25%) done
Select jobs to execute...

[Thu Apr 15 00:21:15 2021]
rule rename:
    input: fastq/SRR13508743_1.fastq.gz, fastq/SRR13508748_1.fastq.gz
    output: fastq/Treatment_1.fastq.gz
    jobid: 3
    wildcards: out=Treatment_1

[Thu Apr 15 00:21:15 2021]
rule rename:
    input: fastq/SRR13508743_1.fastq.gz, fastq/SRR13508748_1.fastq.gz
    output: fastq/Ctrl_1.fastq.gz
    jobid: 1
    wildcards: out=Ctrl_1

Job counts:
	count	jobs
	1	rename
	1
Job counts:
	count	jobs
	1	rename
	1
mv: cannot stat 'fastq/SRR13508743_1.fastq.gz': No such file or directory
[Thu Apr 15 00:21:15 2021]
Error in rule rename:
    jobid: 0
    output: fastq/Treatment_1.fastq.gz

RuleException:
CalledProcessError in line 48 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:
Command 'set -euo pipefail;  set +o pipefail; mv fastq/SRR13508743_1.fastq.gz fastq/Ctrl_1.fastq.gz' returned non-zero exit status 1.
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile", line 48, in __rule_rename
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
Removing output files of failed job rename since they might be corrupted:
fastq/Treatment_1.fastq.gz
[Thu Apr 15 00:21:16 2021]
Finished job 1.
2 of 4 steps (50%) done
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_sra/.snakemake/log/2021-04-14T235436.394836.snakemake.log
```

- Snakefile

```

# This workflow is designed to download fastq files from SRA database. 
# It's possible to perform manually as well (see https://github.com/Mira0507/using_SRA)

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail; source ~/.bashrc; ')
shell.executable('/bin/bash')

rule all: 
    input: 
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])


rule get_fastq: 
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output: 
        expand("fastq/{srr}_{end}.fastq.gz", srr=config['SRA'], end=config['END'])
    params:
        SRA=config['SRA']
    run:
        shell("set +o pipefail; "
              "fastq-dump --split-files {params.SRA} --gzip -O fastq")

rule rename: 
    """
    This rule renames fastq.gz files
    """
    output: 
        "fastq/{out}.fastq.gz"
    input: 
        expand("fastq/{srr}_{end}.fastq.gz", srr=config['SRA'], end=config['END'])
    params: 
        dic=config['SAMPLE'],
        pairs=config['END']
    run:
        for key, value in params.dic.items(): 
            shell("set +o pipefail; "
                  "mv fastq/{key}_1.fastq.gz fastq/{value}_1.fastq.gz")
            if len(params.pairs) == 2: 
                shell("mv fastq/{key}_2.fastq.gz fastq/{value}_2.fastq.gz")

```



- config_single.yaml

```

SAMPLE:
  SRR13508743: Ctrl
  SRR13508748: Treatment




OUTPUT_LIST:
  - Ctrl_1
  - Treatment_1

END: 
  - 1


SRA:
  - SRR13508743
  - SRR13508748



```
