#### 1. Testing: Succeeded

- Snakemake

```
#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail; source ~/.bashrc; ')
shell.executable('/bin/bash')

rule all: 
    input: 
        expand("fastq/{out}.txt", out=config['OUTPUT_LIST'])


rule get_txt: 
    output: 
        "fastq/{out}.txt"
    params: 
        dic=config['SAMPLE'],
        reads=config['END']
    run: 
        shell("set +o pipefail; "
              "touch fastq/SRR13508744_1.txt && " 
              "touch fastq/SRR13508749_1.txt")
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv fastq/{key}_{read}.txt fastq/{value}_{read}.txt")


```

- config/config_single.yaml

```

SAMPLE:
  SRR13508744: Ctrl
  SRR13508749: Treatment




OUTPUT_LIST:
  - Ctrl_1
  - Treatment_1

END: 
  - 1




SRA:
  - SRR13508744
  - SRR13508749



NAME:
  - Ctrl
  - Treatment


```


#### 2. SRA: Failed I

- Snakefile

```

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail; source ~/.bashrc; ')
shell.executable('/bin/bash')

rule all: 
    input: 
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])

rule get_fastq:
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        "fastq/{out}.fastq.gz"
    params:
        dic=config['SAMPLE'],
        reads=config['END'],
        sra=config['SRA']
    run:
        shell("set +o pipefail; "
              "fastq-dump --split-files {params.sra} --gzip")    # Without assigning the output directory (e.g. -O fastq)
        for key, value in params.dic.items(): 
              for read in params.reads: 
                  shell("set +o pipefail; " 
                        "mv {key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz") 



```

- config/config_single.yaml (see above)



- Errors: SRR13508744_1.fastq.gz and SRR13508749_1.fastq.gz files were normally created but Ctrl_1.fastq.gz disappeared at the end

```

(snakemake_sra) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_sra$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	2	get_fastq
	3
Select jobs to execute...

[Thu Apr 15 13:40:40 2021]
rule get_fastq:
    output: fastq/Treatment_1.fastq.gz
    jobid: 2
    wildcards: out=Treatment_1

[Thu Apr 15 13:40:40 2021]
rule get_fastq:
    output: fastq/Ctrl_1.fastq.gz
    jobid: 1
    wildcards: out=Ctrl_1

Job counts:
	count	jobs
	1	get_fastq
	1
Job counts:
	count	jobs
	1	get_fastq
	1
Rejected 22279318 READS because READLEN < 1
Read 22279318 spots for SRR13508744
Written 22279318 spots for SRR13508744
Rejected 22279318 READS because READLEN < 1
Read 22279318 spots for SRR13508744
Written 22279318 spots for SRR13508744
Rejected 23305636 READS because READLEN < 1
Read 23305636 spots for SRR13508749
Written 23305636 spots for SRR13508749
[Thu Apr 15 13:51:29 2021]
Finished job 2.
1 of 3 steps (33%) done
Rejected 23305636 READS because READLEN < 1
Read 23305636 spots for SRR13508749
Written 23305636 spots for SRR13508749
mv: cannot stat 'SRR13508744_1.fastq.gz': No such file or directory    # Where is SRR13508744_1.fastq.gz???????
[Thu Apr 15 13:51:30 2021]
Error in rule get_fastq:
    jobid: 0
    output: fastq/Ctrl_1.fastq.gz

RuleException:
CalledProcessError in line 37 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:
Command 'set -euo pipefail;  set +o pipefail; mv SRR13508744_1.fastq.gz fastq/Ctrl_1.fastq.gz' returned non-zero exit status 1.
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile", line 37, in __rule_get_fastq
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_sra/.snakemake/log/2021-04-15T134039.996766.snakemake.log
```


#### 3. SRA: Failed II 

- Snakefile

```

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')

rule all: 
    input: 
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])

rule get_fastq:
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        "fastq/{out}.fastq.gz"
    params:
        dic=config['SAMPLE'],
        reads=config['END'],
        sra=config['SRA']
    run:
        shell("set +o pipefail; "
              "fastq-dump --split-files {params.sra} --gzip -O fastq")   # The output directory has been defined
        for key, value in params.dic.items(): 
              for read in params.reads: 
                  shell("set +o pipefail; " 
                        "mv fastq/{key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz")

```


- config/config_single.yaml (see above)




- Errors: Same error as the Failed I 

```

(snakemake_sra) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_sra$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	2	get_fastq
	3
Select jobs to execute...

[Thu Apr 15 14:04:01 2021]
rule get_fastq:
    output: fastq/Treatment_1.fastq.gz
    jobid: 2
    wildcards: out=Treatment_1

[Thu Apr 15 14:04:01 2021]
rule get_fastq:
    output: fastq/Ctrl_1.fastq.gz
    jobid: 1
    wildcards: out=Ctrl_1

Job counts:
	count	jobs
	1	get_fastq
	1
Job counts:
	count	jobs
	1	get_fastq
	1
Rejected 22279318 READS because READLEN < 1
Read 22279318 spots for SRR13508744
Written 22279318 spots for SRR13508744
Rejected 22279318 READS because READLEN < 1
Read 22279318 spots for SRR13508744
Written 22279318 spots for SRR13508744
Rejected 23305636 READS because READLEN < 1
Read 23305636 spots for SRR13508749
Written 23305636 spots for SRR13508749
[Thu Apr 15 14:14:46 2021]
Finished job 2.
1 of 3 steps (33%) done
Rejected 23305636 READS because READLEN < 1
Read 23305636 spots for SRR13508749
Written 23305636 spots for SRR13508749
mv: cannot stat 'fastq/SRR13508744_1.fastq.gz': No such file or directory     # Where is fastq/SRR13508744_1.fastq.gz??????
[Thu Apr 15 14:14:47 2021]
Error in rule get_fastq:
    jobid: 0
    output: fastq/Ctrl_1.fastq.gz

RuleException:
CalledProcessError in line 37 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:
Command 'set -euo pipefail;  set +o pipefail; mv fastq/SRR13508744_1.fastq.gz fastq/Ctrl_1.fastq.gz' returned non-zero exit status 1.
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile", line 37, in __rule_get_fastq
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_sra/.snakemake/log/2021-04-15T140401.219466.snakemake.log
```

#### 4. Testing: Failed

- Snakefile

```

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')



rule all: 
    input: 
        expand("fastq/{out}.txt", out=config['OUTPUT_LIST'])


rule get_txt: 
    output: 
        "fastq/{out}.txt"
    params: 
        dic=config['SAMPLE'],
        reads=config['END'],
        sra=config['SRA']
    run: 
        shell("set +o pipefail; "
              "touch fastq/{params.sra}_1.txt")
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv fastq/{key}_{read}.txt fastq/{value}_{read}.txt")
```

- config: see above

- Errors 

```

(snakemake_sra) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_sra$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	2	get_txt
	3
Select jobs to execute...

[Thu Apr 15 14:26:57 2021]
rule get_txt:
    output: fastq/Treatment_1.txt
    jobid: 2
    wildcards: out=Treatment_1

[Thu Apr 15 14:26:57 2021]
rule get_txt:
    output: fastq/Ctrl_1.txt
    jobid: 1
    wildcards: out=Ctrl_1

Job counts:
	count	jobs
	1	get_txt
	1
Job counts:
	count	jobs
	1	get_txt
	1
mv: cannot stat 'fastq/SRR13508744_1.txt': No such file or directory   # Where is fastq/SRR13508744_1.txt??????
[Thu Apr 15 14:26:58 2021]
Error in rule get_txt:
    jobid: 0
    output: fastq/Treatment_1.txt

RuleException:
CalledProcessError in line 37 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:   
Command 'set -euo pipefail;  set +o pipefail; mv fastq/SRR13508744_1.txt fastq/Ctrl_1.txt' returned non-zero exit status 1.
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile", line 37, in __rule_get_txt
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
mv: cannot stat 'fastq/SRR13508744_1.txt': No such file or directory
[Thu Apr 15 14:26:58 2021]
Error in rule get_txt:
    jobid: 0
    output: fastq/Ctrl_1.txt

RuleException:
CalledProcessError in line 37 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:
Command 'set -euo pipefail;  set +o pipefail; mv fastq/SRR13508744_1.txt fastq/Ctrl_1.txt' returned non-zero exit status 1.
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile", line 37, in __rule_get_txt
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_sra/.snakemake/log/2021-04-15T142657.543166.snakemake.log
```



#### 5. Testing: Succeeded

- Snakefile 

```
#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')



rule all: 
    input: 
        expand("fastq/{out}.txt", out=config['OUTPUT_LIST'])


rule get_txt: 
    output: 
        "fastq/{out}.txt"
    params: 
        dic=config['SAMPLE'],
        reads=config['END'],
        sra=config['SRA']
    run: 
        shell("set +o pipefail; "
              "touch fastq/SRR13508744_1.txt && "
              "touch fastq/SRR13508749_1.txt")
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv fastq/{key}_{read}.txt fastq/{value}_{read}.txt")
```

- config: see above


#### 6. Testing: Succeeded

- Snakefile

```
#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')



rule all: 
    input: 
        expand("fastq/{out}.txt", out=config['OUTPUT_LIST'])


rule get_txt: 
    output: 
        "fastq/{out}.txt"
    params: 
        dic=config['SAMPLE'],
        reads=config['END'],
        sra=config['SRA']
    run: 
        shell("set +o pipefail; "
              "touch SRR13508744_1.txt && "
              "touch SRR13508749_1.txt")
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv {key}_{read}.txt fastq/{value}_{read}.txt")
```

- config: see above

#### 7. SRA: Succeeded

- Snakefile 

```
#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')


rule all: 
    input: 
        expand("fastq/{sra}_1.fastq.gz", sra=config['SRA'])

rule get_fastq:
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        "fastq/{sra}_1.fastq.gz"
    params:
        sra=config['SRA']
    run:
        shell("set +o pipefail; "
              "fastq-dump --split-files {params.sra} --gzip -O fastq")  
```

- config: see above

#### 8. SRA: Failed due to cyclic dependency 


- Snakefile 

```

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')


rule all: 
    input: 
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])

rule get_fastq:
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        expand("fastq/{sra}_{end}.fastq.gz", sra=config['SRA'], end=config['END'])
    params:
        sra=config['SRA']
    run:
        shell("set +o pipefail; "
              "fastq-dump --split-files {params.sra} --gzip -O fastq")   

rule rename:
    """
    This rule renames the files
    """
    input: 
        expand("fastq/{sra}_{end}.fastq.gz", sra=config['SRA'], end=config["END"])
    output:
        "fastq/{out}.fastq.gz"
    params:
        dic=config['SAMPLE'],
        reads=config['END']
    run:
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv fastq/{key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz")

```

- Errors

```

(snakemake_sra) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_sra$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	1	get_fastq
	2	rename
	4
Select jobs to execute...

[Thu Apr 15 15:19:50 2021]
rule get_fastq:
    output: fastq/SRR13508744_1.fastq.gz, fastq/SRR13508749_1.fastq.gz
    jobid: 2

Job counts:
	count	jobs
	1	get_fastq
	1
Rejected 22279318 READS because READLEN < 1
Read 22279318 spots for SRR13508744
Written 22279318 spots for SRR13508744
Rejected 23305636 READS because READLEN < 1
Read 23305636 spots for SRR13508749
Written 23305636 spots for SRR13508749
[Thu Apr 15 15:30:37 2021]
Finished job 2.
1 of 4 steps (25%) done
Select jobs to execute...

[Thu Apr 15 15:30:37 2021]
rule rename:
    input: fastq/SRR13508744_1.fastq.gz, fastq/SRR13508749_1.fastq.gz
    output: fastq/Treatment_1.fastq.gz
    jobid: 3
    wildcards: out=Treatment_1

[Thu Apr 15 15:30:37 2021]
rule rename:
    input: fastq/SRR13508744_1.fastq.gz, fastq/SRR13508749_1.fastq.gz
    output: fastq/Ctrl_1.fastq.gz
    jobid: 1
    wildcards: out=Ctrl_1

Job counts:
	count	jobs
	1	rename
	1
Job counts:
	count	jobs
	1	rename
	1
mv: cannot stat 'fastq/SRR13508744_1.fastq.gz': No such file or directory
[Thu Apr 15 15:30:37 2021]
Error in rule rename:
    jobid: 0
    output: fastq/Ctrl_1.fastq.gz

RuleException:
CalledProcessError in line 49 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:
Command 'set -euo pipefail;  set +o pipefail; mv fastq/SRR13508744_1.fastq.gz fastq/Ctrl_1.fastq.gz' returned non-zero exit status 1.
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile", line 49, in __rule_rename
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_sra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
[Thu Apr 15 15:30:37 2021]
Finished job 3.
2 of 4 steps (50%) done
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_sra/.snakemake/log/2021-04-15T151950.534751.snakemake.log
```

#### 9. Testing: Failed

- Snakefile

```

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')


rule all: 
    input: 
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])

rule get_fastq:
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        expand("fastq/{sra}_{end}.fastq.gz", sra=config['SRA'], end=config['END'])
    params:
        sra=config['SRA'],
        end=config['END']
    run:
        for x in params.sra:
            for y in params.end: 
                shell("touch fastq/{x}_{y}.fastq.gz")


rule rename:
    """
    This rule renames the files
    """
    input: 
        expand("fastq/{sra}_{end}.fastq.gz", sra=config['SRA'], end=config["END"])
    output:
        "fastq/{out}.fastq.gz"
    params:
        dic=config['SAMPLE'],
        reads=config['END']
    run:
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv fastq/{key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz")

```

- Errors: Cyclic dependency on rule rename

```

(snakemake_sra) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_sra$ snakemake -j 1
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	1	get_fastq
	2	rename
	4
Select jobs to execute...

[Thu Apr 15 15:39:19 2021]
rule get_fastq:
    output: fastq/SRR13508744_1.fastq.gz, fastq/SRR13508749_1.fastq.gz
    jobid: 2

Job counts:
	count	jobs
	1	get_fastq
	1
[Thu Apr 15 15:39:20 2021]
Finished job 2.
1 of 4 steps (25%) done
Select jobs to execute...

[Thu Apr 15 15:39:20 2021]
rule rename:
    input: fastq/SRR13508744_1.fastq.gz, fastq/SRR13508749_1.fastq.gz
    output: fastq/Ctrl_1.fastq.gz
    jobid: 1
    wildcards: out=Ctrl_1

Job counts:
	count	jobs
	1	rename
	1
[Thu Apr 15 15:39:20 2021]
Finished job 1.
2 of 4 steps (50%) done
Select jobs to execute...

[Thu Apr 15 15:39:20 2021]
rule rename:
    input: fastq/SRR13508744_1.fastq.gz, fastq/SRR13508749_1.fastq.gz
    output: fastq/Treatment_1.fastq.gz
    jobid: 3
    wildcards: out=Treatment_1

CyclicGraphException in line 38 of /home/mira/Documents/programming/Bioinformatics/snakemake_sra/Snakefile:
Cyclic dependency on rule rename.
Removing output files of failed job rename since they might be corrupted:
fastq/Treatment_1.fastq.gz
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_sra/.snakemake/log/2021-04-15T153919.372884.snakemake.log
```

#### 10. SRA: Succeeded

- Snakefile

```
#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################
shell.prefix('set -euo pipefail; ')
shell.executable('/bin/bash')


rule all: 
    input: 
        "fastq/Ctrl_1.fastq.gz",
        "fastq/Treatment_1.fastq.gz"


rule get_fastq: 
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        expand("fastq/{sra}_{end}.fastq.gz", sra=config["SRA"], end=config["END"])
    params:
        sra=config['SRA']
    shell:
        "fastq-dump --split-files {params.sra} --gzip -O fastq"




rule rename: 
    """
    This rule renames fastq.gz files
    """
    input: 
        expand("fastq/{sra}_{end}.fastq.gz", sra=config["SRA"], end=config["END"])
    output:
        "fastq/Ctrl_1.fastq.gz",
        "fastq/Treatment_1.fastq.gz"
    params:
        reads=config['END'],
        dic=config['SAMPLE']
    run:
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv fastq/{key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz")
```


#### 11. SRA: Succeeded


```
#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################
shell.prefix('set -euo pipefail; ')
shell.executable('/bin/bash')


rule all: 
    input: 
        expand("fastq/{name}_{end}.fastq.gz", name=config['NAME'], end=config['END'])


rule get_fastq: 
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        expand("fastq/{sra}_{end}.fastq.gz", sra=config["SRA"], end=config["END"])
    params:
        sra=config['SRA']
    shell:
        "fastq-dump --split-files {params.sra} --gzip -O fastq"




rule rename: 
    """
    This rule renames fastq.gz files
    """
    input: 
        expand("fastq/{sra}_{end}.fastq.gz", sra=config["SRA"], end=config["END"])
    output:
        expand("fastq/{name}_{end}.fastq.gz", name=config['NAME'], end=config['END'])  # This expand() is required to avoid the cyclic dependency
    params:
        reads=config['END'],
        dic=config['SAMPLE']
    run:
        for key, value in params.dic.items(): 
            for read in params.reads: 
                shell("set +o pipefail; "
                  "mv fastq/{key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz")

```


#### 12. SRA: Succeeded!!!!!!!!!!!!


```

#################################### Defined by users #################################
configfile: "config/config_single.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail; ')
shell.executable('/bin/bash')

rule all: 
    input: 
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])

rule get_fastq:
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        expand("fastq/{out}.fastq.gz", out=config['OUTPUT_LIST'])
    params:
        dic=config['SAMPLE'],
        reads=config['END'],
        sra=config['SRA']
    run:
        shell("set +o pipefail; "
              "fastq-dump --split-files {params.sra} --gzip")    # Without assigning the output directory (e.g. -O fastq)
        for key, value in params.dic.items(): 
              for read in params.reads: 
                  shell("set +o pipefail; " 
                        "mv {key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz") 
```
